### ISSUE TITLE
unmanaged race condition during time between shutdown() and closesocket()

### ISSUE THREAD

**Issue by mewalig** [civetweb/civetweb#213]
I noticed a race condition that can occur between shutdown() and closesocket() and accept(). The code I am using is somewhat customized, so my example here is piecemeal. If this isn't sufficient to highlight the issue, please let me know and at some point I'll rebuild an example from the current source.

Here is what I am seeing (on occasion) when I run the server and then issue a large number of concurrent requests against it. I am running on OSX.

```
Thread B executes shutdown() on socket X from within close_socket_gracefully()
Thread A executes accept() from within accept_new_connection(). Apparently, the prior shutdown() call makes it look like socket X is available, so accept() returns socket X
Thread B continues to execute close_socket_gracefully() and executes closesocket() on socket X
Thread A now holds a closed socket X, but thinks that it is still valid and continues to try to process it.
```

Note that when I debugged this using printf output, the output did not reflect the above timing. However, it was clear that the race condition only occurred when the above four steps had all occurred just before it, so I am guessing that my printf() statements were not reflecting the actual sequence of the steps they were reporting on because two printf() calls were being made from two different threads and the printf() calls are not guaranteed to execute together with the instruction they were reporting on.

In any case, one of the connections was getting prematurely dropped every time the above happened (setting a break point at
```
                mg_cry(conn,
                       "%s: setsockopt(SOL_SOCKET SO_LINGER) failed: %s",
```
will stop right after it happens), and I was able to stop it from happening by doing the following:
* add a flag in struct socket:
```
struct socket {
  ...
  char open; /* added */
}
```
* in accept_new_connection(), set so.open if accept() succeeds:
```
        if ((so.sock = accept(listener->sock, &so.rsa.sa, &len))
            == INVALID_SOCKET
            || !(so.open = 1) /* added this line to set so.open */
        ) {
```
* follow every closesocket(xxx.sock) call with "xxx.open = 0;", e.g.:
```
        closesocket(conn->client.sock);
        conn->client.open = 0; /* added */
```
* before executing accept_new_connection(), check if the related listener->open is set:
```
			for (i = 0; i < ctx->num_listening_sockets; i++) {
                          if(!ctx->listening_sockets[i].open) { /* added */
                            ...
                          } /* added */
````

As far as I can tell that fixed the race issue.

**Comment by bel2125** [comment]
Just to be sure I understood it correctly:

Let's assume, one thread is in `close_socket_gracefully` and executes the following line:
`shutdown(conn->client.sock, SHUT_WR);`
Let's assume `conn->client.sock == 42` here, so it's `shutdown(42, SHUT_WR);`
Some lines later it would call
`closesocket(conn->client.sock);`
so the operation is `closesocket(42);`

But just between these two operations another thread calls executing  the `accept_new_connection` function executes the line
`so.sock = accept(listener->sock, &so.rsa.sa, &len)`
and gets 42 as a result?


I think this is not possible, since `shutdown` does not close the socket but just initiates the TCP/IP shutdown sequence. We use `SHUT_WR` here, which means shut down for writing. It is still possible to read from the socket.
From my experience, even after `closesocket`, the socket handle (the value 42) will not be recycled for some time. Probably the operating system still does some cleanup which requires the handle in not recycled immediately.
But I have no experience with OSX / iOS / MacOS .. I don't have any Apple device to test it either.
Still I can not imagine this is any different in OSX - at least I do not think any operating system will return a handle still in use for a not fully closed socket.

**Comment by mewalig** [comment]
Yes, this is what is happening. So accept_new_connection was returning the
same socket twice in a row. I was able to reproduce the issue reliably,
about once per second, until I made the changes I described in the ticket.

It is possible that it was occurring *during* the closesocket() that
immediately followed the shutdown()-- I did not bother to pinpoint the
issue that finely. I just noticed that the issue *only* occurred after the
shutdown() call.

I'll try to gin up an example that occurs using the unmodified code build.
It might be a few days though.
